Upgrade ppp to upstream 2.5.0

icegood opened the pull request:
Report on removed patches:
deleted: 120-debian_ipv6_updown_option.patch: enable_ipv6cp in upstream
deleted: 133-fix_sha1_include.patch: hardly refactored in upstream
deleted: 200-makefile.patch: enable_eaptls=no, with_pcap=no, HAVE_CRYPT_H=1 in configure
deleted: 201-mppe_mppc_1.1.patch: should be inside enable_microsoft_extensions but... https://github.com/ppp-project/ppp/issues/429, discussable...
deleted: 203-opt_flags.patch: fPIC ignored so far
deleted: 300-filter-pcap-includes-lib.patch: done by autotools deleted: 511-pptp_cflags.patch: in main patch for pptp plugin deleted: 600-Revert-pppd-Use-openssl-for-the-DES-instead-of-the-l.patch: with_openssl=no, already in upstream ppp-des.c deleted: 610-pppd_compile_fix.patch: with_static_pcap=yes from patch #310

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

hauke commented on package/network/services/ppp/patches/600-Revert-pppd-Use-openssl-for-the-DES-instead-of-the-l.patch:
```diff
@@ -1,89 +0,0 @@
-For musl and glibc2.27 we can keep linking to crypt; however if we
-switch to glibc 2.28 we will have to link to one of the SSL libraries.
-
-This reverts commit 3c7b86229f7bd2600d74db14b1fe5b3896be3875.
```
Please keep this change to not depend on openssl. We patches glibc to add DES too.

icegood replied:
> Please keep this change to not depend on openssl. We patches glibc to add DES too.

Hi. Could you please help me to understand then the purpose of ppp-des.c in upstream? Doesn't it do the same?

icegood replied:
checked a bit deeper. No changes needed. No openssl dependency present anyway. So, ppp doesn't need either openssl or  patchred glibc

hauke replied:
ppp should not depend on openssl by default. I am not sure how common ppp connections using DES are in the field. If ppp is able to use DES without openssl and this patch now tat would be nice.

Neustradamus replied:
@hauke:
- https://github.com/ppp-project/ppp/pull/362
- https://github.com/enaess/ppp/blob/master/pppd/ppp-des.c

enaess replied:
@icegood just submitted an PR to fix that in 2.5.+. Apologies for that stray openssl include. Technically, you should be able to compile pppd without eaptls or peap features enabled, which allows you to use the built in sha1/md5/md4 and des/ecb functions.

hauke replied:
ok, this should be fine PPP now ships its own DES functionality.

rsalvaterra commented:
Additionally, `chat` seems to be failing to build.

Neustradamus commented:
To follow

icegood commented:
> Additionally, `chat` seems to be failing to build.

fixed. Could you please also help me, how do I  check compilation/installation
of these dependants properly. For now, i check via 'make world' however logically to have something like 
make package/chat/dist or
make package/ppp-multilink/compile

mrkiko commented:
Looking at some git history, definitely worth upgrading.

dchard commented:
FIY, upgrading PPP to 2.5.0 also fixes a PPP netlink issue with LEDs brought to sunlight (but not caused) by Ansuel's latest LED patchset. I already tested it with Ansuel and 2.5.0 fixes the problem.

BKPepe commented on package/network/services/ppp/Makefile:
```diff
@@ -9,20 +9,21 @@ include $(TOPDIR)/rules.mk
 
 PKG_NAME:=ppp
-PKG_RELEASE:=4
+PKG_RELEASE:=5
```
```suggestion
PKG_RELEASE:=1
```

icegood replied:
could you please explain more what strategy of assigning of local version? once merged with upstream then start from 1? Why do we need local  version at all?

icegood replied:
done, but please answer anyway.

icegood replied:
ok, i figure it out...

stklcode replied:
The local or package version indicates this is the N-th version of the upstream sources. It should be incremented when we add changes like patches, build flags, or similar on top.

E.g. upstream version 2.4.9-1 is the first build of upstream 2.4.9
We add a patch to fix something or backport a feature, but still operate on the 2.4.9 upstream sources, then we package 2.4.9-2 and so on...

Bumping the upstream sources we should then reset the package version, s.t. we will produce let’s say 2.5.0-1, our first package based on 2.5.0 upstream. Does not matter if we still have custom patches on top or what the actual upstream changes are, it’s still the first package version that will be built.

enaess commented:
Question for the OpenWRT team. After upgrade to 2.5.0, would you still need a long list of patches for pppd? List in the top of this thread would make me hide ...

icegood commented:
> Question for the OpenWRT team. After upgrade to 2.5.0, would you still need a long list of patches for pppd? List in the top of this thread would make me hide ...

Personally, i removed half of them while merging. The other part should be discussed as they might have some useful features that are still absent in upstream. Let others explain if they know more

icegood committed:
ppp: updated to upstream 2.5.0

Report on removed patches:
deleted: 120-debian_ipv6_updown_option.patch: enable_ipv6cp in upstream
deleted: 133-fix_sha1_include.patch: hardly refactored in upstream
deleted: 200-makefile.patch: enable_eaptls=no, with_pcap=no, HAVE_CRYPT_H=1 in configure
deleted: 201-mppe_mppc_1.1.patch: enable_microsoft_extensions=yes
deleted: 203-opt_flags.patch: fPIC ignored so far
deleted: 300-filter-pcap-includes-lib.patch: done by autotools
deleted: 511-pptp_cflags.patch: in main patch for pptp plugin
deleted: 600-Revert-pppd-Use-openssl-for-the-DES-instead-of-the-l.patch: with_openssl=no, already in upstream ppp-des.c
deleted: 610-pppd_compile_fix.patch: with_static_pcap=yes from patch #310

Signed-off-by: Sergey Ivanov <icegood1980@gmail.com>

wongsyrone commented:
FYI, here's my initial commit on upgrading ppp to the latest upstream commit: https://github.com/wongsyrone/lede-1/commit/5f117f0cedbeb5bb8ad6885521a948d833b6df68

My current tree: https://github.com/wongsyrone/lede-1/tree/master/package/network/services/ppp

I mainly use ppp's PPPoE functionality, with personal changes included, patches prefix with '999-*'.

icegood committed:
ppp: removed patches 140-144 from #12898

Removed these patches as they seem to from upstream(?)
Hasn't dug deep inside

Signed-off-by: Sergey Ivanov <icegood1980@gmail.com>

icegood commented:
> FYI, here's my initial commit on upgrading ppp to the latest upstream commit: [wongsyrone/lede-1@5f117f0](https://github.com/wongsyrone/lede-1/commit/5f117f0cedbeb5bb8ad6885521a948d833b6df68)
> 
> My current tree: https://github.com/wongsyrone/lede-1/tree/master/package/network/services/ppp
> 
> I mainly use ppp's PPPoE functionality, with personal changes included, patches prefix with '999-*'.

what's rationale to have one more?

ikghx commented:
After testing, it is found that this update will cause pptpd and rp-pppoe compilation to fail.

partial error log:

/home/test/x86/openwrt/build_dir/target-x86_64_musl/pptpd-1.4.0=pptpd-1.4.0 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -I.. -I../../include -fPIC pptpd-logwtmp.c -lutil
pptpd-logwtmp.c:17:23: error: 'VERSION' undeclared here (not in a function)
   17 | char pppd_version[] = VERSION;
      |                       ^~~~~~~
pptpd-logwtmp.c:19:31: error: 'PATH_MAX' undeclared here (not in a function)
   19 | static char pptpd_original_ip[PATH_MAX+1];
      |                               ^~~~~~~~
pptpd-logwtmp.c:19:13: error: size of array 'pptpd_original_ip' has non-integer type
   19 | static char pptpd_original_ip[PATH_MAX+1];
      |             ^~~~~~~~~~~~~~~~~
pptpd-logwtmp.c:22:8: error: unknown type name 'option_t'
   22 | static option_t options[] = {
      |        ^~~~~~~~
pptpd-logwtmp.c:23:3: warning: braces around scalar initializer
   23 |   { "pptpd-original-ip", o_string, pptpd_original_ip,
      |   ^

zdz2019 commented:
After testing, it is found that this update will cause pptpd and rp-pppoe compilation to fail.

partial error log:

/home/test/x86/openwrt/build_dir/target-x86_64_musl/pptpd-1.4.0=pptpd-1.4.0 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -I.. -I../../include -fPIC pptpd-logwtmp.c -lutil
pptpd-logwtmp.c:17:23: error: 'VERSION' undeclared here (not in a function)
17 | char pppd_version[] = VERSION;
| ^~~~~~~
pptpd-logwtmp.c:19:31: error: 'PATH_MAX' undeclared here (not in a function)
19 | static char pptpd_original_ip[PATH_MAX+1];
| ^~~~~~~~
pptpd-logwtmp.c:19:13: error: size of array 'pptpd_original_ip' has non-integer type
19 | static char pptpd_original_ip[PATH_MAX+1];
| ^~~~~~~~~~~~~~~~~
pptpd-logwtmp.c:22:8: error: unknown type name 'option_t'
22 | static option_t options[] = {
| ^~~~~~~~
pptpd-logwtmp.c:23:3: warning: braces around scalar initializer
23 | { "pptpd-original-ip", o_string, pptpd_original_ip,
| ^

Neustradamus commented:
@zdz2019: About RP-PPPoE, it is needed to have the last one too:
- https://dianne.skoll.ca/projects/rp-pppoe/
- https://github.com/dfskoll/rp-pppoe/tags
- https://dianne.skoll.ca/pipermail/rp-pppoe/2023q2/thread.html

Badly, about "pptpd" Poptop, the development does not exist since a long time :/
- https://sourceforge.net/projects/poptop/

enaess commented:
I looked at PopTop pptpd source and found no really generic way to mask for the 2.5.0 upgrade. Project is pretty much dead as no changes has been made there for years so I gave up... 

Having that said, perhaps the OpenWRT project should consider dropping poptop altogether. Apple did, was it back in 2018?, due to security concerns with the protocol. People should use l2tp or sstp instead.

enaess commented:
@zdz2019 
For the record, pptpd was released on October 23, 2013. That is almost 10 years ago
https://sourceforge.net/projects/poptop/files/pptpd/

I did not submit patches to the project as I considered it stale / out of date.

enaess commented:
Also:
https://security.stackexchange.com/questions/45509/are-there-any-known-vulnerabilities-in-pptp-vpns-when-configured-properly

icegood commented:
> pptpd

openwrt has own patches to bring pptpd up: 500, 510 . However it wasn't tested

icegood commented:
> @zdz2019: About RP-PPPoE, it is needed to have the last one too:
> 
> * https://dianne.skoll.ca/projects/rp-pppoe/
> * https://github.com/dfskoll/rp-pppoe/tags
> * https://dianne.skoll.ca/pipermail/rp-pppoe/2023q2/thread.html
> 
> Badly, about "pptpd" Poptop, the development does not exist since a long time :/
> 
> * https://sourceforge.net/projects/poptop/

pptp answer for this

icegood commented:
Guys, let's concentrate on 90% of more useful test cases. I finally was able to build and test against my router.  For now, given PR doesn't work with my provider against main branch of the rest of openwrt. 
Syslog: 
`
Sat Jun 17 17:29:16 2023 daemon.notice netifd: Interface 'wan' is now down
Sat Jun 17 17:29:16 2023 daemon.notice netifd: Interface 'wan' is setting up now
Sat Jun 17 17:29:16 2023 daemon.info pppd[3844]: Plugin pppoe.so loaded.
Sat Jun 17 17:29:16 2023 daemon.info pppd[3844]: PPPoE plugin from pppd 2.5.0
Sat Jun 17 17:29:16 2023 daemon.notice pppd[3844]: pppd 2.5.0 started by root, uid 0
Sat Jun 17 17:29:17 2023 daemon.warn odhcpd[1316]: No default route present, overriding ra_lifetime!
Sat Jun 17 17:29:31 2023 daemon.warn pppd[3844]: Timeout waiting for PADO packets
Sat Jun 17 17:29:31 2023 daemon.err pppd[3844]: Unable to complete PPPoE Discovery phase 1
Sat Jun 17 17:29:31 2023 daemon.info pppd[3844]: Exit.
Sat Jun 17 17:29:31 2023 daemon.notice netifd: Interface 'wan' is now down
Sat Jun 17 17:29:31 2023 daemon.notice netifd: Interface 'wan' is setting up now
Sat Jun 17 17:29:32 2023 daemon.info pppd[3946]: Plugin pppoe.so loaded.
Sat Jun 17 17:29:32 2023 daemon.info pppd[3946]: PPPoE plugin from pppd 2.5.0
Sat Jun 17 17:29:32 2023 daemon.notice pppd[3946]: pppd 2.5.0 started by root, uid 0
Sat Jun 17 17:29:32 2023 daemon.warn odhcpd[1316]: No default route present, overriding ra_lifetime!
Sat Jun 17 17:29:47 2023 daemon.warn pppd[3946]: Timeout waiting for PADO packets
Sat Jun 17 17:29:47 2023 daemon.err pppd[3946]: Unable to complete PPPoE Discovery phase 1
Sat Jun 17 17:29:47 2023 daemon.info pppd[3946]: Exit.

`
And this is strange since accordingly to tcpdump pad0 is delivered right after initiation (in a less than 10ms), but nothing happens next after it. 
The only note is that first initiation doesn't have the answer. Does it break the state machine? Full tcpdump on a wwan interface is:
https://drive.google.com/file/d/1A4N688W6T1krGC0I-tsDkwy5Waq2HFvI/view?usp=drive_link
Router to test: tl-wa801nd-v5

Neustradamus commented:
@enaess, @icegood: Oh yes, @wongsyrone has updated 500/510 patches here:
- https://github.com/wongsyrone/lede-1/tree/master/package/network/services/ppp/patches

icegood commented:
> Guys, let's concentrate on 90% of more useful test cases. I finally was able to build and test against my router. For now, given PR doesn't work with my provider against main branch of the rest of openwrt. Syslog: ` Sat Jun 17 17:29:16 2023 daemon.notice netifd: Interface 'wan' is now down Sat Jun 17 17:29:16 2023 daemon.notice netifd: Interface 'wan' is setting up now Sat Jun 17 17:29:16 2023 daemon.info pppd[3844]: Plugin pppoe.so loaded. Sat Jun 17 17:29:16 2023 daemon.info pppd[3844]: PPPoE plugin from pppd 2.5.0 Sat Jun 17 17:29:16 2023 daemon.notice pppd[3844]: pppd 2.5.0 started by root, uid 0 Sat Jun 17 17:29:17 2023 daemon.warn odhcpd[1316]: No default route present, overriding ra_lifetime! Sat Jun 17 17:29:31 2023 daemon.warn pppd[3844]: Timeout waiting for PADO packets Sat Jun 17 17:29:31 2023 daemon.err pppd[3844]: Unable to complete PPPoE Discovery phase 1 Sat Jun 17 17:29:31 2023 daemon.info pppd[3844]: Exit. Sat Jun 17 17:29:31 2023 daemon.notice netifd: Interface 'wan' is now down Sat Jun 17 17:29:31 2023 daemon.notice netifd: Interface 'wan' is setting up now Sat Jun 17 17:29:32 2023 daemon.info pppd[3946]: Plugin pppoe.so loaded. Sat Jun 17 17:29:32 2023 daemon.info pppd[3946]: PPPoE plugin from pppd 2.5.0 Sat Jun 17 17:29:32 2023 daemon.notice pppd[3946]: pppd 2.5.0 started by root, uid 0 Sat Jun 17 17:29:32 2023 daemon.warn odhcpd[1316]: No default route present, overriding ra_lifetime! Sat Jun 17 17:29:47 2023 daemon.warn pppd[3946]: Timeout waiting for PADO packets Sat Jun 17 17:29:47 2023 daemon.err pppd[3946]: Unable to complete PPPoE Discovery phase 1 Sat Jun 17 17:29:47 2023 daemon.info pppd[3946]: Exit.
> 
> ` And this is strange since accordingly to tcpdump pad0 is delivered right after initiation (in a less than 10ms), but nothing happens next after it. The only note is that first initiation doesn't have the answer. Does it break the state machine? Full tcpdump on a wwan interface is: https://drive.google.com/file/d/1A4N688W6T1krGC0I-tsDkwy5Waq2HFvI/view?usp=drive_link Router to test: tl-wa801nd-v5
> 
> @enaess , do we have high level diagramm on ppp? And sequence ones at least with happy use cases?
no help? no answer?

enaess commented:
@icegood While I have worked on pppd, my working knowledge of pppoe may not be my strongest suits, and there is no good diagram of how this should work. I think you probably should post the configuration of pppd w/pppoe as you are trying to run it and see if one of us would be able to reproduce it locally without openwrt. I have tested 2.5.0 prior to the release, and pppoe was working.

Generally, the plugin (pppoe.so) is creating a pppox socket and connects it with the real interface eth0 or "wan" if that is what you called it. The kernel should route PADO packets via the pppox socket and pppd/pppoe plugin should be able to write back to respond to the PADO message. The plugin does expect the interface to be named to certain pattern, so maybe check that first, and also make sure to enable "debug" in your configuration.

Ansuel commented:
@icegood did you by chance tried to report this to the repo itself? Smells like a regression of the new version...

Neustradamus commented:
Dear all OpenWrt devs,

@paulusmack will prepare a 2.5.1 release build and ask some questions:
- https://github.com/ppp-project/ppp/issues/460

Current upstream is good?
Some patches need to be added in upstream?

Thanks in advance.

ptpt52 commented:
mark to follow

ptpt52 commented:
a patch may fix the build for pptpd
```diff
diff --git a/bcrelay.c b/bcrelay.c
index 3fa8e00..850e27e 100644
--- a/bcrelay.c
+++ b/bcrelay.c
@@ -435,7 +435,7 @@ static void mainloop(int argc, char **argv)
   struct timeval time_2_wait;
   static struct ifsnr old_ifsnr[MAXIF+1]; // Old iflist to socket fd's mapping list
   static struct ifsnr cur_ifsnr[MAXIF+1]; // Current iflist to socket fd's mapping list
-  unsigned char buf[1518];
+  unsigned char buf[2048];
   char *logstr = empty;
 
   no_discifs_cntr = MAX_NODISCOVER_IFS;
diff --git a/plugins/pptpd-logwtmp.c b/plugins/pptpd-logwtmp.c
index ac5ecc2..d7573d5 100644
--- a/plugins/pptpd-logwtmp.c
+++ b/plugins/pptpd-logwtmp.c
@@ -13,6 +13,9 @@
 #include <utmp.h>
 #include <string.h>
 #include <pppd/pppd.h>
+#include <pppd/options.h>
+#include <linux/limits.h>
+#include "patchlevel.h"
 
 char pppd_version[] = VERSION;
 
@@ -28,7 +31,7 @@ static option_t options[] = {
   { NULL }
 };
 
-static char *reduce(char *user)
+static const char *reduce(const char *user)
 {
   char *sep;
   if (!pptpd_logwtmp_strip_domain) return user;
@@ -42,25 +45,25 @@ static char *reduce(char *user)
 
 static void ip_up(void *opaque, int arg)
 {
-  char *user = reduce(peer_authname);
-  if (debug)
-    notice("pptpd-logwtmp.so ip-up %s %s %s", ifname, user, 
+  const char *user = reduce(ppp_peer_authname(NULL, 0));
+  if (debug_on())
+    notice("pptpd-logwtmp.so ip-up %s %s %s", ppp_ifname(), user,
 	   pptpd_original_ip);
-  logwtmp(ifname, user, pptpd_original_ip);
+  //logwtmp(ppp_ifname(), user, pptpd_original_ip);
 }
 
 static void ip_down(void *opaque, int arg)
 {
-  if (debug) 
-    notice("pptpd-logwtmp.so ip-down %s", ifname);
-  logwtmp(ifname, "", "");
+  if (debug_on())
+    notice("pptpd-logwtmp.so ip-down %s", ppp_ifname());
+  //logwtmp(ppp_ifname(), "", "");
 }
 
 void plugin_init(void)
 {
-  add_options(options);
-  add_notifier(&ip_up_notifier, ip_up, NULL);
-  add_notifier(&ip_down_notifier, ip_down, NULL);
-  if (debug) 
+  ppp_add_options(options);
+  ppp_add_notify(NF_IP_UP, ip_up, NULL);
+  ppp_add_notify(NF_IP_DOWN, ip_down, NULL);
+  if (debug_on())
     notice("pptpd-logwtmp: $Version$");
 }
```

curtdept commented:
@Neustradamus this waiting on 2.5.1?

Neustradamus commented:
@curtdept: I do not know when @paulusmack will create the 2.5.1 release and I do not know what is the @icegood situation and this progress :/

curtdept commented:
@Neustradamus @hauke I'm in the process of finding out, but I think this thing is blocking a glibc 2.40 bump as well as they took out crypto almost completely

Neustradamus commented:
PPP 2.5.1 has been released (2024-09-18)
- https://github.com/ppp-project/ppp/releases/tag/ppp-2.5.1

Neustradamus commented:
Dear all, @nbd168,

Can you check for a PPP 2.5.1 update in OpenWrt?
- https://github.com/ppp-project/ppp/issues/460
- https://github.com/ppp-project/ppp/releases/tag/ppp-2.5.1

It is important to have all improvements/bugfixes before next branch.

Thanks in advance.

hauke commented on package/network/services/ppp/patches/201-mppe_mppc_1.1.patch:
```diff
@@ -1,1518 +0,0 @@
-This is a forward ported version of ppp-2.4.3-mppe-mppc-1.1.patch.gz found on
-http://mppe-mppc.alphacron.de/ .
-
-Signed-off-by: Jo-Philipp Wich <jo@mein.io>
```
The MPPC functionality is not supported in upstream ppp, MPPE is supported in upstream and can be activated using `enable_microsoft_extensions` like it is done in this PR. 

Do we need MPPC support? I think it is ok to remove it.

hauke commented:
How well was this PR tested? Which scenarios are known to work and which one still fail?

curtdept commented:
> How well was this PR tested? Which scenarios are known to work and which one still fail?

Sparsely, we all waiting for the 2.5.1 bump for a while 2.5.0 had regressions

DragonBluep commented:
A testing branch for ppp 2.5.1: https://github.com/DragonBluep/openwrt/tree/testing

hauke commented:
> A testing branch for ppp 2.5.1: https://github.com/DragonBluep/openwrt/tree/testing

Thank you for taking care of this. 
I hope it is ok to drop `package/network/services/ppp/patches/201-mppe_mppc_1.1.patch` @jow- what do you think? Do we need MPPC? I think MPPE is already supported upstream. 

Could you add a link to the upstream patch in the description of the ppp 2.5.0 update when you found the patch or something similar upstream. For the build and buildsystem fixes it is not needed, but got the others it would be nice.

nasbdh9 commented:
> A testing branch for ppp 2.5.1: https://github.com/DragonBluep/openwrt/tree/testing

```
make[5]: Entering directory '/home/nasbdh9/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-mediatek_mt7622/ppp-default/ppp-2.5.1.git-2024-09-18/pppd'
aarch64-openwrt-linux-musl-gcc -DHAVE_CONFIG_H -I. -I../pppd/plugins/pppoe  -DSYSCONFDIR=\"/etc\" -DPPPD_RUNTIME_DIR='"/var/run/pppd"' -DPPPD_LOGFILE_DIR='"/var/log/ppp"' -DSYSTEM_CA_PATH='"/etc/ssl/certs"'   -DPPPD_PLUGIN_DIR='"/usr/lib/pppd/2.5.1"' -I/usr/include/security    -I/home/nasbdh9/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-14.2.0_musl/usr/include -I/home/nasbdh9/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-14.2.0_musl/include -I/home/nasbdh9/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-14.2.0_musl/include/fortify   -O2 -pipe -mcpu=cortex-a53 -ftree-vectorize -ftree-slp-vectorize -fno-caller-saves -fno-plt -fhonour-copts -fmacro-prefix-map=/home/nasbdh9/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-mediatek_mt7622/ppp-default/ppp-2.5.1.git-2024-09-18=ppp-2.5.1.git-2024-09-18 -ffunction-sections -fdata-sections -flto=auto -fno-fat-lto-objects -Wformat -Werror=format-security -DPIC -fPIC -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro  -c -o pppd-pcap_pcc.o `test -f 'pcap_pcc.c' || echo './'`pcap_pcc.c
pcap_pcc.c: In function 'pcap_pre_compiled':
pcap_pcc.c:16:8: error: implicit declaration of function 'ppp_option_error' [-Wimplicit-function-declaration]
   16 |        ppp_option_error("error opening precompiled active-filter '%s': %s",
      |        ^~~~~~~~~~~~~~~~
make[5]: *** [Makefile:1419: pppd-pcap_pcc.o] Error 1
```

nasbdh9 commented:
https://github.com/nasbdh9/openwrt/commits/ppp-2.5.1/
All ppp packages should compile fine, tested pppoe, are welcome to test others.
I can open a new PR if needed.

DragonBluep commented:
> > A testing branch for ppp 2.5.1: https://github.com/DragonBluep/openwrt/tree/testing
> 
> ```
> make[5]: Entering directory '/home/nasbdh9/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-mediatek_mt7622/ppp-default/ppp-2.5.1.git-2024-09-18/pppd'
> aarch64-openwrt-linux-musl-gcc -DHAVE_CONFIG_H -I. -I../pppd/plugins/pppoe  -DSYSCONFDIR=\"/etc\" -DPPPD_RUNTIME_DIR='"/var/run/pppd"' -DPPPD_LOGFILE_DIR='"/var/log/ppp"' -DSYSTEM_CA_PATH='"/etc/ssl/certs"'   -DPPPD_PLUGIN_DIR='"/usr/lib/pppd/2.5.1"' -I/usr/include/security    -I/home/nasbdh9/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-14.2.0_musl/usr/include -I/home/nasbdh9/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-14.2.0_musl/include -I/home/nasbdh9/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-14.2.0_musl/include/fortify   -O2 -pipe -mcpu=cortex-a53 -ftree-vectorize -ftree-slp-vectorize -fno-caller-saves -fno-plt -fhonour-copts -fmacro-prefix-map=/home/nasbdh9/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-mediatek_mt7622/ppp-default/ppp-2.5.1.git-2024-09-18=ppp-2.5.1.git-2024-09-18 -ffunction-sections -fdata-sections -flto=auto -fno-fat-lto-objects -Wformat -Werror=format-security -DPIC -fPIC -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro  -c -o pppd-pcap_pcc.o `test -f 'pcap_pcc.c' || echo './'`pcap_pcc.c
> pcap_pcc.c: In function 'pcap_pre_compiled':
> pcap_pcc.c:16:8: error: implicit declaration of function 'ppp_option_error' [-Wimplicit-function-declaration]
>    16 |        ppp_option_error("error opening precompiled active-filter '%s': %s",
>       |        ^~~~~~~~~~~~~~~~
> make[5]: *** [Makefile:1419: pppd-pcap_pcc.o] Error 1
> ```

Fixed. Strange, GCC13 doesn't complain about it.

> > A testing branch for ppp 2.5.1: https://github.com/DragonBluep/openwrt/tree/testing
> 
> Thank you for taking care of this. I hope it is ok to drop `package/network/services/ppp/patches/201-mppe_mppc_1.1.patch` @jow- what do you think? Do we need MPPC? I think MPPE is already supported upstream.
> 
> Could you add a link to the upstream patch in the description of the ppp 2.5.0 update when you found the patch or something similar upstream. For the build and buildsystem fixes it is not needed, but got the others it would be nice.

If we can get some positive feedback here, I'll update commit message later.


> Guys, let's concentrate on 90% of more useful test cases. I finally was able to build and test against my router. For now, given PR doesn't work with my provider against main branch of the rest of openwrt. Syslog: ` Sat Jun 17 17:29:16 2023 daemon.notice netifd: Interface 'wan' is now down Sat Jun 17 17:29:16 2023 daemon.notice netifd: Interface 'wan' is setting up now Sat Jun 17 17:29:16 2023 daemon.info pppd[3844]: Plugin pppoe.so loaded. Sat Jun 17 17:29:16 2023 daemon.info pppd[3844]: PPPoE plugin from pppd 2.5.0 Sat Jun 17 17:29:16 2023 daemon.notice pppd[3844]: pppd 2.5.0 started by root, uid 0 Sat Jun 17 17:29:17 2023 daemon.warn odhcpd[1316]: No default route present, overriding ra_lifetime! Sat Jun 17 17:29:31 2023 daemon.warn pppd[3844]: Timeout waiting for PADO packets Sat Jun 17 17:29:31 2023 daemon.err pppd[3844]: Unable to complete PPPoE Discovery phase 1 Sat Jun 17 17:29:31 2023 daemon.info pppd[3844]: Exit. Sat Jun 17 17:29:31 2023 daemon.notice netifd: Interface 'wan' is now down Sat Jun 17 17:29:31 2023 daemon.notice netifd: Interface 'wan' is setting up now Sat Jun 17 17:29:32 2023 daemon.info pppd[3946]: Plugin pppoe.so loaded. Sat Jun 17 17:29:32 2023 daemon.info pppd[3946]: PPPoE plugin from pppd 2.5.0 Sat Jun 17 17:29:32 2023 daemon.notice pppd[3946]: pppd 2.5.0 started by root, uid 0 Sat Jun 17 17:29:32 2023 daemon.warn odhcpd[1316]: No default route present, overriding ra_lifetime! Sat Jun 17 17:29:47 2023 daemon.warn pppd[3946]: Timeout waiting for PADO packets Sat Jun 17 17:29:47 2023 daemon.err pppd[3946]: Unable to complete PPPoE Discovery phase 1 Sat Jun 17 17:29:47 2023 daemon.info pppd[3946]: Exit.
> 
> ` And this is strange since accordingly to tcpdump pad0 is delivered right after initiation (in a less than 10ms), but nothing happens next after it. The only note is that first initiation doesn't have the answer. Does it break the state machine? Full tcpdump on a wwan interface is: https://drive.google.com/file/d/1A4N688W6T1krGC0I-tsDkwy5Waq2HFvI/view?usp=drive_link Router to test: tl-wa801nd-v5
> 
> @enaess , do we have high level diagramm on ppp? And sequence ones at least with happy use cases?

As for this issue, I suggest removing patch `205-no_exponential_timeout.patch`. This patch doesn't make sense for me. And I think this issue is not a regression. There are already some reports in the forum.
https://forum.openwrt.org/t/pppoe-unknown-error-user-request/48765

nasbdh9 commented:
```
crypto.c: In function 'PPP_crypto_error':
crypto.c:178:11: error: implicit declaration of function 'vsnprintf' [-Wimplicit-function-declaration]
  178 |     off = vsnprintf(buf, len, fmt, args);
      |           ^~~~~~~~~
crypto.c:41:1: note: include '<stdio.h>' or provide a declaration of 'vsnprintf'
   40 | #include "crypto-priv.h"
  +++ |+#include <stdio.h>
   41 |
crypto.c:178:26: warning: 'vsnprintf' argument 2 type is 'int' where 'long unsigned int' is expected in a call to built-in function declared without prototype [-Wbuiltin-declaration-mismatch]
  178 |     off = vsnprintf(buf, len, fmt, args);
      |                          ^~~
<built-in>: note: built-in 'vsnprintf' declared here
make[5]: *** [Makefile:1062: libppp_crypto_la-crypto.lo] Error 1
```
fix：
https://github.com/nasbdh9/openwrt/commit/a76cddfed2196b0960a132f4fbde6a2b388ccf3e#diff-e2c6702c11faa2a20bc705434c0808f4eb0147c08e62cd50a2a9ca4ab3de104d

nasbdh9 commented:
```
install -m0755 ./files/lib/netifd/ppp-down /home/nasbdh9/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-mediatek_mt7622/ppp-default/ppp-2.5.1.git-2024-09-18/ipkg-aarch64_cortex-a53/ppp/lib/netifd/
find /home/nasbdh9/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-mediatek_mt7622/ppp-default/ppp-2.5.1.git-2024-09-18/ipkg-aarch64_cortex-a53/ppp -name 'CVS' -o -name '.svn' -o -name '.#*' -o -name '*~'| xargs -r rm -rf
Package ppp is missing dependencies for the following libraries:
libpam.so.0
make[2]: *** [Makefile:336: /home/nasbdh9/openwrt/bin/packages/aarch64_cortex-a53/base/ppp_2.5.1.git-2024-09-18-r1_aarch64_cortex-a53.ipk] Error 1
```
dependencies need to be updated to meet arm requirements fix:
https://github.com/nasbdh9/openwrt/commit/a76cddfed2196b0960a132f4fbde6a2b388ccf3e#diff-b6a245380d75179adbb4eb8cb7509279ef46f9edba267a2ce11affa462b50282R45

hauke commented:
Removing `205-no_exponential_timeout.patch` is fine with me. Please do it in a separate commit so we can revert it if we run into problems.

If possible do not add extra dependencies like libpam and libpacp to ppp. ppp is part of the default image and we would like to keep it small.

If it is needed for some useful use cases add a `ppp-tiny` and a `ppp-full` package like it is done for `ip`. 

I would like to get the new ppp version into OpenWrt. Multiple commits are looking like they are fixing potential security problems. This will make it also easier for us to report problems upstream.

I assume that we can not fully test it, because there are probably a lot of different (old) devices out there we want to establish ppp connections to. All the different devices Telkos bought in the last 20 years.

Neustradamus commented:
What do you think about PPP 2.5.1?
- https://github.com/openwrt/openwrt/pull/16605

hauke commented:
The ppp update was merged in #16605.
